File: examples/canonical-analytics/go/internal/analytics/incidents.go
package analytics

import (
	"analytics-workbench-go/internal/data"
	"analytics-workbench-go/internal/filters"
)

type IncidentSeverity struct {
	Severities     []string
	Counts         []int
	MTTRBySeverity map[string]float64
}

func AggregateIncidents(incidents []data.Incident, services []data.Service, f filters.FilterState) IncidentSeverity {
	validServices := make(map[string]bool)
	if len(services) > 0 {
		for _, s := range services {
			validServices[s.Name] = true
		}
	} else {
		for _, i := range incidents {
			validServices[i.Service] = true
		}
	}

	var filtered []data.Incident
	for _, i := range incidents {
		if !validServices[i.Service] {
			continue
		}
		if !f.ContainsService(i.Service) {
			continue
		}
		if !f.ContainsEnvironment(i.Environment) {
			continue
		}
		if !f.ContainsSeverity(i.Severity) {
			continue
		}
		t := parseDate(i.Timestamp)
		if !f.InDateRange(t) {
			continue
		}
		filtered = append(filtered, i)
	}

	if len(filtered) == 0 {
		return IncidentSeverity{
			MTTRBySeverity: make(map[string]float64),
		}
	}

	order := []string{"sev1", "sev2", "sev3", "sev4"}
	
	counts := make(map[string]int)
	mttrSum := make(map[string]int)
	
	for _, i := range filtered {
		counts[i.Severity]++
		mttrSum[i.Severity] += i.MttrMins
	}

	var presentSeverities []string
	countList := make([]int, 0)
	mttrMap := make(map[string]float64)

	for _, sev := range order {
		if c, ok := counts[sev]; ok && c > 0 {
			presentSeverities = append(presentSeverities, sev)
			countList = append(countList, c)
			mttrMap[sev] = float64(mttrSum[sev]) / float64(c)
		}
	}

	return IncidentSeverity{
		Severities:     presentSeverities,
		Counts:         countList,
		MTTRBySeverity: mttrMap,
	}
}

File: examples/canonical-analytics/rust/src/analytics/incidents.rs
use super::{match_date, match_multi, IncidentSeverity};
use crate::data::models::{Incident, Service};
use crate::filters::FilterState;
use std::collections::{HashMap, HashSet};

pub fn aggregate_incident_severity(
    incidents: &[Incident],
    services: &[Service],
    filters: &FilterState,
) -> IncidentSeverity {
    let valid_services: HashSet<String> = services.iter().map(|s| s.name.clone()).collect();

    let mut counts_map: HashMap<String, i64> = HashMap::new();
    let mut mttr_sum: HashMap<String, f64> = HashMap::new();

    for i in incidents {
        if !match_date(&i.timestamp, &filters.date_from, &filters.date_to) {
            continue;
        }
        if !match_multi(&i.environment, &filters.environment) {
            continue;
        }
        if !match_multi(&i.severity, &filters.severity) {
            continue;
        }
        if !valid_services.contains(&i.service) {
            continue;
        }
        if !match_multi(&i.service, &filters.services) {
            continue;
        }

        *counts_map.entry(i.severity.clone()).or_insert(0) += 1;
        *mttr_sum.entry(i.severity.clone()).or_insert(0.0) += i.mttr_mins;
    }

    if counts_map.is_empty() {
        return IncidentSeverity {
            severities: Vec::new(),
            counts: Vec::new(),
            mttr_by_severity: HashMap::new(),
        };
    }

    let order = vec!["critical", "high", "medium", "low", "sev1", "sev2", "sev3", "sev4"];
    let mut severities = Vec::new();
    let mut counts = Vec::new();
    let mut mttr_by_severity = HashMap::new();

    for sev in order {
        if let Some(&count) = counts_map.get(sev) {
            let sum = *mttr_sum.get(sev).unwrap();
            severities.push(sev.to_string());
            counts.push(count);
            mttr_by_severity.insert(sev.to_string(), sum / count as f64);
        }
    }

    IncidentSeverity {
        severities,
        counts,
        mttr_by_severity,
    }
}
